### PR TITLE
Catch changed URLError message from `ftp_open`

### LINKED ISSUE #14091: v5.2rc1 `_download_file_from_source` does not catch `URLError` on Python 3.10.8 + 3.11.0
### Description
On latest 3.10 and 3.11 Pythons `test_ftp_tls_auto` fails on `download_file(url)` with
>           raise URLError(exp) from exp
> E           urllib.error.URLError: <urlopen error 530 Anonymous sessions must use encryption.>

### Expected behavior
Call to `astropy.utils.data._download_file_from_source(url)` with the default `ftp_tls=None` should revert to `ftp_tls=True` after raising a permission exception.

### Actual behavior
Exception handling has changed in 3.10.8 and 3.11.0 with python/cpython#24938, now raising with the original error message instead of the "ftplib error" expected by `_download_file_from_source`.

Can be fixed by changing https://github.com/astropy/astropy/blob/252055ef3b9d6be3dd0e7ab3e7b70e1f28b80421/astropy/utils/data.py#L1287
to
```python
            if str(e.reason).startswith(("ftp error: error_perm", "530 Anonymous")):
```
As I understand the change in Python, the previous wrapper would also replace the original args with just the error class, so the above fix would not be quite as general as for Python <= 3.10.7; to cover all `error_perm` (= class 500) exceptions, we might simply check for "5" in the second case – within the context of `URLError` this might still be specific enough?

### Steps to Reproduce
```python
astropy.test(package='utils', args='-vk test_ftp_tls_auto', remote_data=True)
```

### System Details
macOS-12.6.1-arm64-arm-64bit
Python 3.10.8 (main, Oct 12 2022, 21:27:37) [Clang 14.0.0 (clang-1400.0.29.201)]
Python 3.11.0 (main, Oct 24 2022, 23:26:06) [Clang 14.0.0 (clang-1400.0.29.201)]
Numpy 1.23.5
pyerfa 2.0.0.1
astropy 5.2rc1
Scipy 1.9.3
Matplotlib 3.5.1

### PR DESCRIPTION
**Pull request by dhomeier**
### Description
This pull request is to address a change in raising exceptions in `ftp_open` in latest Pythons to handle them in `_download_file_from_source`.

Fixes #14091

This is using the more general check discussed there to catch any permission (type 5NN) exceptions.
In released Python 3.11.0 the exception has been changed to 
`raise URLError(f'ftp error: {exp}') from exp`
vs. 
`raise URLError(exp) from exp`
in 3.10.8, 3.12a2 and main, so that may change again in the next 3.11 release – catching all 3 versions here.
Also giving `test_ftp_tls_auto` a new try here.

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label. Codestyle issues can be fixed by the [bot](https://docs.astropy.org/en/latest/development/workflow/development_workflow.html#pre-commit).
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by dhomeier** [comment]
`test_iers_a_dl_mirror` timeout seems unrelated.
`test_ftp_tls_auto` passing on py311-test-devdeps; since that is allowed to fail and the only env with remote-data=any, might be safe to leave it in.

**Review by mhvk — COMMENTED** [review]
Thanks! Looks good except perhaps to keep skipping the IERS A download test.

File: astropy/utils/tests/test_data.py
Comment (bsipocz) [reply astropy/utils/tests/test_data.py:2141]: pytest-rerunfailure maybe an answer for real flaky stuff, assuming the timescale of the flakyness is not hours or days but for consecutive runs. 
See astroquery for example usage.

Comment (mhvk) [reply astropy/utils/tests/test_data.py:2141]: I actually am really not so sure how meaningful it is to spam the IERS A download site with repeated requests for what are not small files. This might make more sense on monthly cron.

Comment (mhvk) [reply astropy/utils/tests/test_data.py:2141]: p.s. But for *this* PR I would suggest to just leave the `skipif` in place and get your nice solution in!

Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2141]: Yes, the download takes not-insignificant time, so putting it into weekly or monthly cron would make more sense – but the only one using remote data is py310-test-alldeps-predeps in cron_daily, it seems. Perhaps there also is a smaller file we could use?

Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2141]: > – but the only one using remote data is `py310-test-alldeps-predeps` in cron_daily, it seems.

Except, is it really? It has `toxposargs: --remote-data=any` (ci_workflows have `posargs: --remote-data=foo`), but is not using any remote-data flags and the [last daily](https://github.com/astropy/astropy/actions/runs/3598502712/jobs/6061324996#step:9:198) was running with 
```
Using Astropy options: remote_data: none.

ARCH_ON_CI: undefined
IS_CRON: undefined
```
@pllim do you know if `os.environ.get("CI")` should be set there?

Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: I think "CI" env var is set up by Actions. It should always be true in CI setting.

FWIW, @saimn also put in the `hugemem` flag but we currently do not enable it on CI.

https://github.com/astropy/astropy/blob/252055ef3b9d6be3dd0e7ab3e7b70e1f28b80421/astropy/io/ascii/tests/test_c_reader.py#L1157

Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: Ah, yes, here is the documentation from Actions: https://docs.github.com/en/actions/learn-github-actions/environment-variables#default-environment-variables

Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: I propose we add the skipif back in and open up a follow-up issue. I don't think we need to really test this in backport branch, but only going forward in main?

I agree a cron to actually test this is good.

Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: Might wanna squash too? 😸

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:62]: Now that I think about it... Is this a bug? 😱 
```suggestion
CI = os.environ.get("CI", "false") == "true"
```

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:63]:
```suggestion
```

File: astropy/utils/tests/test_data.py
Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2141]: > Might wanna squash too? 😸

Just replaced the other commit 😏 – it is skipped in the workflows now, although conftest does not report the `IS_CRON` value correctly, and it won't run in cron_daily while remote_data are not enabled.
I wonder if the variables set in `env` are passed through the OpenAstronomy workflow; and in particular found it lists `toxargs` and `posargs` as inputs, but not `toxposargs`. But that will be another issue indeed!

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:2142]: This is enough, right? You wouldn't set this var or CI locally anyway.
```suggestion
@pytest.mark.skipif(os.environ.get("IS_CRON", "false") == "false", reason="Flaky/too much external traffic for CI")
```

File: astropy/utils/tests/test_data.py
Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: Oh, right... OpenAstronomy does not pass custom env var... See https://github.com/astropy/astropy/pull/14070#discussion_r1035351968 . But I reverted the weekly cron to not use it in https://github.com/astropy/astropy/pull/14077 . So theoretically, this check means this test would run on weekly cron only, so we might need to change this?

https://github.com/astropy/astropy/blob/252055ef3b9d6be3dd0e7ab3e7b70e1f28b80421/.github/workflows/ci_cron_weekly.yml#L42-L43

File: astropy/utils/tests/test_data.py
Comment (dhomeier) [reply astropy/utils/tests/test_data.py:62]: That would be cleaner for sure, though should not change the outcome. 😺

File: astropy/utils/tests/test_data.py
Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2142]: Yes, should have seen I could set this similarly to the Windows skips, but got a bit confused about the `"CI"` passed as string – why was that anyway.
I'd nonetheless combine both however, so this would still run in interactive tests.

File: astropy/utils/tests/test_data.py
Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2141]: Yes, and weekly would probably be enough – I did not quite understand why remote was disabled there, but good for daily anyway.

File: astropy/utils/tests/test_data.py
Comment (pllim) [reply astropy/utils/tests/test_data.py:62]: No, but my head hurts less. 😆

File: astropy/utils/tests/test_data.py
Comment (pllim) [inline astropy/utils/tests/test_data.py:2141]: Checking for CI would skip even the cron job, right? It is still technically CI because it is run in Actions.

```suggestion
@pytest.mark.skipif(os.environ.get("IS_CRON", "false") == "false",
```

Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2141]: Thus the `and` – we don't want to disable it for everything else, do we (would never ever have found it in the RC2-tests then...)?

**Comment by mhvk** [comment]
Let's keep the changelog entry!

File: astropy/utils/tests/test_data.py
Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: Hmm I guess we can leave it in... maybe someone could have defined their own `IS_CRON` on their local machine for whatever reason.

**Review by pllim — COMMENTED** [review]
You need to rename "bug" to "bugfix" for your change log fragment. Also please rebase and squash the commits. Thanks!

I also pinged Erik just because he wrote the original feature. But I'll approve anyway after the change log stuff is fixed. Thanks!

File: astropy/utils/tests/test_data.py
Comment (pllim) [reply astropy/utils/tests/test_data.py:2141]: @bsipocz , rerunfailure does look promising for some remote data tests we have here, but that is a battle for another day. Thanks for the suggestion!

File: astropy/utils/tests/test_data.py
Comment (dhomeier) [reply astropy/utils/tests/test_data.py:2141]: Which would not matter unless they also made a habit of defining `CI` locally... 🤔

**Comment by dhomeier** [comment]
> You need to rename "bug" to "bugfix" for your change log fragment. Also please rebase and squash the commits. Thanks!

Doh, of course. Do you want it squashed to two commits or a single one?

**Comment by pllim** [comment]
pre-commit.ci autofix

**Comment by dhomeier** [comment]
> pre-commit.ci autofix

Is there anything to be done about these? Did not quite understand where the error was.

**Comment by pllim** [comment]
> Is there anything to be done about these?

I think if you like do local precommit hooks or something, you can catch it before the bot. cc @nstarman and @WilliamJamieson

**Comment by nstarman** [comment]
If you cd to the repo and do `pre-commit install`. Now it'll run pre-commit on the diff before any commit is made.

**Comment by nstarman** [comment]
Many things have autofix. Pre-commit will "fail" but fix the code. Then you just re-stage the changed files and re-commit and it passes.

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

**Comment by dhomeier** [comment]
> If you cd to the repo and do `pre-commit install`. Now it'll run pre-commit on the diff before any commit is made.

You mean if I have pre-commit setup locally for that? OK, maybe will do that eventually just to save the extra CI cycle, but I thought the point was that black was applied fully automatically with no need for further action (or at least as little as possible).

**Comment by pllim** [comment]
That was what the autofix command do if it fails the CI.

**Comment by dhomeier** [comment]
OK, so it still takes the autofix comment to actually make and commit the changes; thanks.
@WilliamJamieson is it also possible to mark a commit so it will skip the CI _except_ pre-commit.ci (so the full CI would only be run once after the final style fixes have been applied)?

**Comment by pllim** [comment]
Re: skip -- If it runs on Actions like the CI, then it is bound by the same directives as the CI, which means we cannot just run it but not the CI without putting in even more customization.

**Comment by pllim** [comment]
Unless... it doesn't and doesn't respect `[ci skip]`? 🤷

**Comment by mhvk** [comment]
Ideally, on every push the pre-commit stuff would run before anything else, and would be required for any tests to run, so that no regular tests are ever run on code that has errors that can be fixed in precommit.

**Comment by pllim** [comment]
@mhvk , are you proposing that we run precommit job first in CI and have other jobs rely on its success?

**Comment by pllim** [comment]
Actually... not sure how that is possible with the current set up. The bot runs as a separately installed app, not part of our own workflow that we can control.

**Comment by dhomeier** [comment]
Yes, I first thought it should go under `initial_checks` just as codestyle, but the reasoning in https://github.com/astropy/astropy/pull/13553#issuecomment-1215564956 was that it should not be applied fully automatically in any case, so no conflicts with the contributor's branch (without pre-commit applied) would arise. I actually appreciate it that way, so I won't have to deal with black until everything is (about) ready to merge.
According to that, the bot should respect all the `[ci skip...]` directives, plus its own set of `[pre-commit.ci...]` ones. I was basically wondering if there was the complementary version to that, i.e. skipping only the CI on Actions, but not the pre-commit one.

**Pull request merged**
